### DISCUSSION TITLE
Command to loop a single track

### DISCUSSION THREAD

**Discussion by ibmerlin2** [Makuna/DFMiniMp3#75] [Q&A] [answered]
Is there a command that would let you loop a single track over and over? I am creating a music box and would like to continue playing a track while the lid is open.

**Comment by Makuna** [comment, marked as answer]
The [Wiki](https://github.com/Makuna/DFMiniMp3/wiki/API-Reference#void-loopglobaltrackuint16_t-globaltrack) is your friend.

**Reply by ibmerlin2** [reply]
Thank you, I should have searched further before asking.

**Reply by Makuna** [reply]
If you look, there are several APIs, try them until you find one that works for your needs.  Note that due to differences in the production of the chips, some don't support all the API but they provide no way for the user or library to know what they do or don't support.

**Reply by massimo2021** [reply]
I'm trying to convert a RFID juukebox that works with the DFPlayer version MH2024K-24SS to new chip MH2024K-16SS (no Checksum)
The last obstacle is the repetition of the single track.
Please, can you help me?

```
DFMiniMp3<SoftwareSerial, Mp3Notify> mp3(secondarySerial);
|
|
 //PLAY SONG
     mp3.playMp3FolderTrack(number.toInt());
     isPlaying = true;

    //mp3.setRepeatPlayCurrentTrack(true); //NO
    //DfMp3_PlaybackMode_Repeat; //NO
    //mp3.loopGlobalTrack(number.toInt()); yes , but only by date 
    //mp3.setRepeatPlayCurrentTrack(1, repeat);  // NO
    // mp3.setRepeatPlayCurrentTrack(repeat); // NO
```

**Reply by Makuna** [reply]
Sounds the like  MH2024K-16SS just isn't very compatible.  You can do it "manually" by using the callback.
```
class Mp3Notify  {
public:
// removed other callback methods for brevity
  
  static void OnPlayFinished(DfMp3& mp3, DfMp3_PlaySources source, uint16_t track)  {
     if (track == g_repeatTrack) {
       mp3.playMp3FolderTrack(g_repeatTrack);
     }
  }
};
```

**Reply by massimo2021** [reply]
Thanks a lot.

I'll try the code but first I have also to re-check all the connections, because it gives me an error

```
Com Error 129
volume 0
```

(my configuration Arduino Uno + MFRC522 + DFPlayer mini MH2024K-16SS)

**Reply by Makuna** [reply]
see DfMp3_Error, 129 is 0x81, which is an expected data was not received and the serial comms timed out.  

You didn't say which call gave you this error.

**Reply by massimo2021** [reply]
Oops.    You are right.
```
void setup() {
  Serial.begin(9600);
  Serial.setTimeout(500);
  Serial.println("initializing...");
  secondarySerial.begin(9600);
  mp3.begin(9600);
  Serial.setTimeout(500);
  
  SPI.begin();                                                  // Init SPI bus
  mfrc522.PCD_Init();                                      // Init MFRC522 card

  pinMode(playPauseButton, INPUT_PULLUP);
  pinMode(shuffleButton, INPUT_PULLUP);

  Serial.println(F("Initializing DFPlayer ... (May take 3~5 seconds)"));

  uint16_t volume = mp3.getVolume();
  Serial.print("volume ");
  Serial.println(volume);
  mp3.setVolume(20);
```

**Reply by Makuna** [reply]
There is no need to set the baud rate for secondarySerial, its managed within the MP3 class.  The timeout for the mp3 serial is set to 10 seconds, as some calls do take a while.  But again, I have no experience with this newer chip that seems to be so incompatible with the previous versions it might need longer.

**Comment by seisfeld** [comment]
> mp3.setRepeatPlayCurrentTrack(1, repeat); // NO
> mp3.setRepeatPlayCurrentTrack(repeat); // NO

Did you actually try:

`mp3.setRepeatPlayCurrentTrack(TRUE);`

**Reply by massimo2021** [reply]
I tried.
Sometimes it works, sometimes not.
At this point I think it's a hardware problem.
Maybe a badly welded pin of the DFPlayer
I will prove in the coming days

Thanks :-)

**Reply by Schallbert** [reply]
Hey Massimo,
I had a very similar problem in my project, sometimes it just wouldn't play the next track but sometimes it worked. I figured out that in my case the callback logic (onTrackFinished or so) wouldn't work so that my program didn't get the signal about a finished track. I solved this in my project by reading the isPlaying pin of the DFmini. If it stopped when playback is complete, I made it play the next track, see here how I did this: https://github.com/Schallbert/Tonuino_alternativeFirmware/blob/9824cebf345bf940115b620d3268e0478c6c6648/lib/Mp3/Mp3Play/Mp3Play_implementation.cpp#L67
Instead of playing the next track, you could just call the function with the same track again and it should work all the time.

**Reply by massimo2021** [reply]
Thanks.
I'm not at your level with C ++.
But I understand the concept. I'll try to apply it to my code.
By the way, I saw the project
https://schallbert.github.io/projects-software/tonuino/
Really very nice and original. I've never seen a cards reader as complete and accurate.

**Reply by massimo2021** [reply]
_Finally!_
After 8 days of tests and experiments, now the code works.

The MP3 player connects to the RFID and can read and repeat the chosen track.

I tested the project by shaking cables, turning on and turning off many times and reseting: the DFPlayer continues to work :-)

All I had to do was add some  delay(1000);

__ Thank you all for the support __

********************************************

```
//---------------------------------------- GET NUMBER AND PLAY THE SONG

    byte buffer2[18];
    block = 1;
    len = 18;

    status = mfrc522.PCD_Authenticate(MFRC522::PICC_CMD_MF_AUTH_KEY_A, 1, &key, &(mfrc522.uid)); //line 834
    if (status != MFRC522::STATUS_OK) {
      Serial.print(F("Authentication failed: "));
      Serial.println(mfrc522.GetStatusCodeName(status));
      return;
    }

    status = mfrc522.MIFARE_Read(block, buffer2, &len);
    if (status != MFRC522::STATUS_OK) {
      Serial.print(F("Reading failed: "));
      Serial.println(mfrc522.GetStatusCodeName(status));
      return;
    }


    //PRINT NUMBER
    String number = "";

    for (uint8_t i = 0; i < 16; i++)
    {
      number += (char)buffer2[i];
    }
    number.trim();
    Serial.print(number);

    //PLAY SONG

    
    mp3.playMp3FolderTrack(number.toInt());

delay(1000);

    mp3.setRepeatPlayCurrentTrack(true);
    
    isPlaying = true;
    
   //----------------------------------------

delay(1000);

    Serial.println(F("\n**End Reading**\n"));

delay(1000);

    mfrc522.PICC_HaltA();
    mfrc522.PCD_StopCrypto1();
```
